server/server.go

```diff
@@ -0,0 +1,377 @@
+}
+
+type addFileResult struct {
+	Node format.Node
+	Err  error
+}
+
+func addFile(ctx context.Context, peer *ipfslite.Peer, addParams *pb.AddParams, reader io.Reader, ch chan addFileResult) {
+	defer close(ch)
+	params := &ipfslite.AddParams{
+		Layout:    addParams.GetLayout(),
+		Chunker:   addParams.GetChunker(),
+		RawLeaves: addParams.GetRawLeaves(),
+		Hidden:    addParams.GetHidden(),
+		Shard:     addParams.GetShared(),
+		NoCopy:    addParams.GetNoCopy(),
+		HashFun:   addParams.GetHashFun(),
+	}
+	node, err := peer.AddFile(ctx, reader, params)
+	if err != nil {
+		ch <- addFileResult{Err: err}
+		return
+	}
+	ch <- addFileResult{Node: node}
+}
+
+func (s *ipfsLiteServer) AddFile(srv pb.IpfsLite_AddFileServer) error {
+	req, err := srv.Recv()
+	if err != nil {
+		return err
+	}
+	var addParams *pb.AddParams
+	switch payload := req.GetPayload().(type) {
+	case *pb.AddFileRequest_AddParams:
+		addParams = payload.AddParams
+	default:
+		return fmt.Errorf("expexted AddParams for AddFileRequest.Payload but got %T", payload)
+	}
+
+	reader, writer := io.Pipe()
+
+	addFileChannel := make(chan addFileResult)
+	go addFile(srv.Context(), s.peer, addParams, reader, addFileChannel)
+
+	for {
+		req, err := srv.Recv()
+		if err == io.EOF {
+			writer.Close()
+			break
+		} else if err != nil {
+			writer.CloseWithError(err)
+			break
+		}
+		switch payload := req.GetPayload().(type) {
+		case *pb.AddFileRequest_Chunk:
+			_, writeErr := writer.Write(payload.Chunk)
+			if writeErr != nil {
+				return writeErr
+			}
+		default:
+			return fmt.Errorf("expected Chunk for AddFileRequest.Payload but got %T", payload)
+		}
+	}
+
+	addFileResult := <-addFileChannel
+	if addFileResult.Err != nil {
+		return addFileResult.Err
+	}
+
+	respNode, err := nodeToPbNode(addFileResult.Node)
+	if err != nil {
+		return err
+	}
+
+	return srv.SendAndClose(&pb.AddFileResponse{Node: respNode})
+}
+
+func (s *ipfsLiteServer) GetFile(req *pb.GetFileRequest, srv pb.IpfsLite_GetFileServer) error {
+	cid, err := cid.Decode(req.GetCid())
+	if err != nil {
+		return err
+	}
+
+	reader, err := s.peer.GetFile(srv.Context(), cid)
+	if err != nil {
+		return err
+	}
+
+	buffer := make([]byte, getFileChunkSize)
+	for {
+		size, err := reader.Read(buffer)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		srv.Send(&pb.GetFileResponse{Chunk: buffer[:size]})
+		if err == io.EOF {
+			return nil
+		}
+	}
+}
+
+func (s *ipfsLiteServer) HasBlock(ctx context.Context, req *pb.HasBlockRequest) (*pb.HasBlockResponse, error) {
+	cid, err := cid.Decode(req.GetCid())
+	if err != nil {
+		return nil, err
+	}
+	hasBlock, err := s.peer.HasBlock(cid)
+	if err != nil {
+		return nil, err
+	}
+	return &pb.HasBlockResponse{HasBlock: hasBlock}, nil
+}
+
+func (s *ipfsLiteServer) AddNode(ctx context.Context, req *pb.AddNodeRequest) (*pb.AddNodeResponse, error) {
+	block, err := pbBlockToBlock(req.GetBlock())
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := format.Decode(block)
```

Comment:
So I get my block back above (created without specifying a cid) and this call to `Decode` fails with the error:
```
failed to add node: rpc error: code = Unknown desc = failed to decode Protocol Buffers: incorrectly formatted merkledag node: unmarshal failed. unexpected EOF```
If I change the test to specify the cid, this all works no problem.

Reply:
It's strange that it thinks the data is Protobuf data when I encoded it with cbor.
<img width="800" alt="Screen Shot 2019-11-01 at 5 23 49 PM" src="https://user-images.githubusercontent.com/528969/68061865-9089f900-fccc-11e9-9125-6d58a84adf0e.png">
That screen shot is me stepping into the decoding process. You can see that the detected decoder is for Protobuf data.